### PR TITLE
HEEDLS-288 Ignore archived tutorials in data services

### PR DESCRIPTION
**Pull request by benjimarshall**
### Changes
Ignore archived tutorials in data services by adding filters
to `WHERE` clauses in various queries.

### Testing
* Test that archived tutorials are ignored (both in lists of tutorials, and when aggregating scores over tutorials).
* Rewrite the tests comparing to the old stored procedures because these services now diverge from the stored procedures (they would include archived tutorials).

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs [hunk 1]
```diff
@@ -502,63 +502,146 @@ public void Get_course_content_should_only_return_sections_with_tutorials_in_cus
...
+                .GetCourseContent(candidateId, customisationId)?
+                .Sections
+                .Select(section => section.Id);
+
+            // Then
+            sectionIdsInCourseContent.Should().Contain(103);
+        }
+
+        [Test]
+        public void Get_course_content_uses_overrideTutorialMins_when_calculating_duration()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs:571]: There are more changes to these tests than I expected. Is that due to missing out cases like this the first time? Or have these tests been moved from somewhere else? 🙂

File: DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs [hunk 2]
```diff
@@ -502,63 +502,146 @@ public void Get_course_content_should_only_return_sections_with_tutorials_in_cus
...
+            const int candidateId = 11;
+            const int customisationId = 15937;
+
+            // When
+            var result = courseContentService.GetCourseContent(candidateId, customisationId);
+
+            // Then
+            result.Should().NotBeNull();
+            result!.Sections.First(section => section.Id == 392)
+                .PercentComplete.Should().Be(100.0 * 3 / 14);
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs:598]: Quite a few magic numbers here. Could you extract them to some well named variables please? 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs [hunk 3]
```diff
@@ -398,7 +398,7 @@ public void Get_diagnostic_content_should_not_return_tutorials_where_diagAssess_
         }
 
         [Test]
-        public void Get_diagnostic_content_should_not_return_tutorials_where_archived_date_is_null()
+        public void Get_diagnostic_content_should_not_return_archived_tutorials()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:401]: Might have small merge conflict with @kwunkiuma

File: DigitalLearningSolutions.Data.Tests/Services/SectionContentServiceTests.cs [hunk 4]
```diff
@@ -331,36 +344,41 @@ public void Get_section_content_should_order_by_tutorial_id_when_shared_order_by
...
-        [TestCase(254480, 24224, 101, 1, null)]
-        public void Get_section_content_should_have_same_tutorials_as_stored_procedure(
-            int candidateId,
-            int customisationId,
-            int centreId,
-            int sectionId,
-            int? progressId
-        )
+        [Test]
+        public void Get_section_content_should_sort_using_orderByNumber()
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/SectionContentServiceTests.cs:348]: This seems unrelated to the current change. Has this test just been moved?

File: DigitalLearningSolutions.Data/Services/SectionContentService.cs [hunk 5]
```diff
@@ -82,9 +82,10 @@ GROUP BY
                     WHERE
                         CustomisationTutorials.CustomisationID = @customisationId
                         AND Sections.SectionID = @sectionId
-                        AND (Sections.ArchivedDate IS NULL)
+                        AND Sections.ArchivedDate IS NULL
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/SectionContentService.cs:85]: What about the id for the next section, are we checking that isn't archived?

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 6]
```diff
@@ -176,7 +178,8 @@ FROM CustomisationTutorials
                          AND Sections.SectionID = @sectionId
                          AND Tutorials.TutorialId = @tutorialId
                          AND Customisations.Active = 1
-                         AND CustomisationTutorials.Status = 1;",
+                         AND CustomisationTutorials.Status = 1
+                         AND Tutorials.ArchivedDate IS NULL;",
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:182]: Are we checking the section isn't archived?

**Review by FridaTveit — COMMENTED** [review]
This contains changes to a lot of queries so is a bit hard to review. It would have been useful to split it up to avoid conflicts. Have you checked with @kwunkiuma and @SChandarana that there are no conflicts? A HLD would also have been useful 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 7]
```diff
@@ -54,6 +54,7 @@ FROM Tutorials
 
                          LEFT JOIN Tutorials AS NextTutorials
                          ON NextTutorials.SectionID = Tutorials.SectionID
+                            AND NextTutorials.ArchivedDate IS NULL
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:57]: No need to check archived date of the tutorial in the join below? Or of the section in the two joins on the sections table?

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 8]
```diff
@@ -147,6 +148,7 @@ FROM Tutorials
                      AND Tutorials.TutorialID = @tutorialId
                      AND Customisations.Active = 1
                      AND CustomisationTutorials.Status = 1
+                     AND Tutorials.ArchivedDate IS NULL
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:151]: What about if the section is achived?

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs [hunk 9]
```diff
@@ -206,7 +209,8 @@ FROM CustomisationTutorials
                          AND Sections.SectionID = @sectionId
                          AND Tutorials.TutorialId = @tutorialId
                          AND Customisations.Active = 1
-                         AND CustomisationTutorials.Status = 1;",
+                         AND CustomisationTutorials.Status = 1
+                         AND Tutorials.ArchivedDate IS NULL;",
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/TutorialContentService.cs:213]: What about if the section is archived?

**Review by FridaTveit — COMMENTED** [review]
A few additional comments. I intentionally haven't been thinking about whether there are any more queries we should be checking/changing, because if that is the case then I think we should do that in a new PR after getting this merged. Will have a look through the queries tomorrow 🙂

**Comment by benjimarshall** [comment]
@FridaTveit I totally agree that this should have been split up over a couple of PRs, each doing a few of the services. I still can do that (maybe after sorting this round of review comments) if that would be useful. 

The actual SQL changes are all fairly minor, but because we are now diverging from the old stored procedure's behaviour I thought it was best to extract the different cases that the "service is like old stored procedure" tests were covering into named tests that don't rely on the old stored procedures. I went through each of the cases for each of these tests and tried to work out which cases parts of the query they were testing to then put into tests, though I admit I got a bit carried away and there might have been a better way of doing this. Sorry for not checking with you! 🙂

File: DigitalLearningSolutions.Data.Tests/Services/SectionContentServiceTests.cs (hunk 4 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/SectionContentServiceTests.cs:348]: This was one of the things being checked using the stored procedure `Get_section_content_should_have_same_tutorials_as_stored_procedure`, so I added this when I removed that test 🙂

File: DigitalLearningSolutions.Data/Services/SectionContentService.cs (hunk 5 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/SectionContentService.cs:85]: This line is just some brackets getting removed for consistency I think. The query isn't getting the next section ID here, because I haven't merged in Shiv's latest changes. I'll add those in soon 🙂

**Comment by FridaTveit** [comment]
Yes, I know the SQL changes are small but making sure they're correct is still not straight forward. Similarly for the test changes. So for next time I would recommend splitting it up 🙂 

For right now, it depends how much work you think it would be to split it up. If it won't be too much work then I think it would be useful. It would make it easier to see how the test changes match up with SQL query changes, and means the more straight forward queries could get merged quickly. But if it's going to be difficult then it's okay to keep it as it is, just remember for next time 🙂

File: DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs (hunk 2 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/CourseContentServiceTests.cs:598]: Yep! Fixed in an upcoming PR 🙂

File: DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/DiagnosticAssessmentServiceTests.cs:401]: This didn't turn out be too bad, but thanks for the heads up 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 6 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs:182]: No, added this change in #154 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 7 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs:57]: Yes we need to check the archived dates for that `NextSectionTutorial`. The first section join is the current section, so is checked with the change in the previous markup. The next section is now being checked, thank you for that! All done in #154 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 9 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs:213]: Yep, I hadn't thought of that! Fixed in #154 🙂

File: DigitalLearningSolutions.Data/Services/TutorialContentService.cs (hunk 8 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/TutorialContentService.cs:151]: Yep, I hadn't thought of that! Fixed in #154 🙂

File: DigitalLearningSolutions.Data/Services/SectionContentService.cs (hunk 5 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/SectionContentService.cs:85]: Added in #157 🙂

**Comment by benjimarshall** [comment]
All review comments have been resolved in subsequent PRs, and PRs #154, #156 and #157 make all of the changes proposed in this PR
